### PR TITLE
Feature/amendments to json endpoint

### PR DESCRIPTION
**Pull request by franmoore05**
### What

Amended /json endpoint so a geography can now be fully flexed.  Added a new function to do this as the multivariates work in future may need the previous one created so we can keep it.

### How to review

Ensure the tests pass and the code makes sense.

### Who can review

Anyone.

### CODE REVIEW

File: api/dataset.go [hunk 1]
```diff
@@ -94,8 +94,56 @@ func (api *API) getDatasetJSON(ctx context.Context, r *http.Request, params *dat
 	return response, nil
 }
 
+func (api *API) validateGeography(ctx context.Context, r *http.Request, params *datasetParams, datasetRequest cantabular.StaticDatasetQueryRequest) ([]string, *[]cantabular.Filter, error) {
```
Comment (ChrysmOre) [inline api/dataset.go:97]: Never any need to make a pointer to a slice as a slice essentially acts like a pointer in itself (including being able to equal nil)

File: api/dataset.go [hunk 2]
```diff
@@ -94,8 +94,56 @@ func (api *API) getDatasetJSON(ctx context.Context, r *http.Request, params *dat
...
+		// Check the area code is in the area type
+		for _, a := range geographyOptions {
+			cReq := cantabular.GetAreaRequest{
+				Dataset:  params.basedOn,
+				Variable: geography,
+				Category: a,
+			}
+
+			_, err := api.ctblr.GetArea(ctx, cReq)
+
```
Comment (ChrysmOre) [inline api/dataset.go:128]: better not to leave blank lines before checking errors

File: api/dataset.go [hunk 3]
```diff
@@ -94,8 +94,56 @@ func (api *API) getDatasetJSON(ctx context.Context, r *http.Request, params *dat
...
+	if len(geographyOptions) > 0 {
+		// Check the area code is in the area type
+		for _, a := range geographyOptions {
+			cReq := cantabular.GetAreaRequest{
+				Dataset:  params.basedOn,
+				Variable: geography,
+				Category: a,
+			}
+
+			_, err := api.ctblr.GetArea(ctx, cReq)
```
Comment (ChrysmOre) [inline api/dataset.go:127]: Isn't there a `cantabular.GetAreas` function that could query all at once?

File: api/dataset.go [hunk 4]
```diff
@@ -156,22 +204,23 @@ func (api *API) getGeographyFilters(r *http.Request, params *datasetParams) ([]c
...
 }
 
 func (api *API) toGetDatasetJsonResponse(params *datasetParams, query *cantabular.StaticDatasetQuery) (*GetDatasetJSONResponse, error) {
 	var dimensions []DatasetJSONDimension
 
 	for _, dimension := range query.Dataset.Table.Dimensions {
-		if _, ok := params.options[dimension.Variable.Name]; !ok {
-			return nil, errors.New("dimension mismatch")
-		}
+		// if _, ok := params.options[dimension.Variable.Name]; !ok {
```
Comment (ChrysmOre) [inline api/dataset.go:214]: Meant to be still commented out or removed?

File: api/api.go [hunk 5]
```diff
@@ -108,5 +108,7 @@ func (api *API) enablePrivateEndpoints() {
 	r.Put("/filter-outputs/{id}", api.putFilterOutput)
 	r.Post("/filter-outputs/{id}/events", api.addFilterOutputEvent)
 
+	r.Get("/datasets/{dataset_id}/editions/{edition}/versions/{version}/json", api.getDatasetJSONHandler)
```
Comment (ChrysmOre) [inline api/api.go:111]: If the endpoint is now meant to be available to publishing then some secrets need to be updated for dp-api-router

**Review by ChrysmOre — COMMENTED** [review]
Some comments

File: api/dataset.go (hunk 3 above)
Comment (j-s-rawat) [reply api/dataset.go:127]: This is a test to validate that all geography options are valid. There is a possibility of getting all the areas i.e. GetAreas, and then check if geography options are valid in  list returned by GetAreas, but GetAreas is based on pagination and will have to implement pagination logic to retrieve all the list. 
So unless geography options is a big list and cantabular.GetArea is inefficient, there  is no harm in the way it is currently implemented.

File: api/api.go (hunk 5 above)
Comment (j-s-rawat) [reply api/api.go:111]: Create a [PR](https://github.com/ONSdigital/dp-configs/pull/1131) to address this issue

**Review by thomasridd — DISMISSED** [review]
This is a priority ticket and all PR comments have been addressed. I'm approving it to open up the opportunity to merge

**Review by ChrysmOre — APPROVED** [review]

**Pull request merged**